drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
+        self.write_header(msgtype, payload_size);
+        self.write_crc(payload_size)
+    }
+
+    /// Serialize a request from a MsgType and Lease
+    pub fn from_lease(
+        &mut self,
+        msgtype: MsgType,
+        source: Leased<R, [u8]>,
+    ) -> Result<VerifiedTxMsg, SprotError> {
+        if source.len() > PAYLOAD_SIZE_MAX {
+            return Err(SprotError::Oversize);
         }
+
+        let dest = &mut self.buf[HEADER_SIZE..][..source.len()];
+        source
+            .read_range(0..source.len(), dest)
+            .map_err(|_| SprotError::TaskRestart)?;
+
+        self.write_header(msgtype, source.len());
+        Ok(self.write_crc(source.len()))
     }
-}
 
-impl From<UpdateRspKind> for u8 {
-    fn from(kind: UpdateRspKind) -> Self {
-        match kind {
-            UpdateRspKind::Ok => UpdateRspKind::Ok as u8,
-            UpdateRspKind::Value => UpdateRspKind::Value as u8,
-            UpdateRspKind::Error => UpdateRspKind::Error as u8,
-            _ => UpdateRspKind::Unknown as u8,
+    /// Serialize a request from a buffer with an already written payload
+    pub fn from_existing(
+        &mut self,
+        msgtype: MsgType,
+        payload_size: usize,
+    ) -> Result<VerifiedTxMsg, SprotError> {
+        if payload_size > PAYLOAD_SIZE_MAX {
+            return Err(SprotError::Oversize);
         }
+        self.write_header(msgtype, payload_size);
+        Ok(self.write_crc(payload_size))
     }
-}
 
-#[derive(
-    Copy, Clone, Serialize, Deserialize, SerializedSize, Eq, PartialEq,
-)]
-#[repr(C, packed)]
-pub struct UpdateRspHeader {
-    pub kind: UpdateRspKind,
-    pub value: u32, // unused, value, or error code depending on kind.
-}
+    fn write_header(&mut self, msgtype: MsgType, payload_size: usize) {
+        let _ = MsgHeader::new_v1(msgtype, payload_size)
+            .unwrap_lite()
+            .serialize(&mut self.buf[..])
+            .unwrap_lite();
+    }
 
-// Hubpack doesn't serialize Option or Result but that's
-// what we want.
-// Map the various Update api results to a signle struct
-// that hubpack can handle.
-// match Result<(), UpdateError> {
-//  Ok(()) => { new(None, None) },
-//  Err(err) => { new(None, err.into()) },
-// match Result<u32, UpdateError> {
-//  Ok(block_size) => { new(Some(block_size), None) },
-//  Err(err) => { new(None, err.into()) },
-impl UpdateRspHeader {
-    pub fn new(arg: Option<u32>, err: Option<u32>) -> Self {
-        match err {
-            None => match arg {
-                None => Self {
-                    kind: UpdateRspKind::Ok,
-                    value: 0,
-                },
-                Some(value) => Self {
-                    kind: UpdateRspKind::Value,
-                    value,
-                },
-            },
-            Some(error) => Self {
-                kind: UpdateRspKind::Error,
-                value: error,
-            },
-        }
+    fn write_crc(&mut self, payload_size: usize) -> VerifiedTxMsg {
```

Comment:
I'd be tempted to merge this with `write_header`; otherwise, you're not sure that the `payload_size` argument necessarily matches what was written to the header.

(of course, we're doing it correctly at the moment, but it seems like a cleaner API to do them simultaneously)

Reply:
At that point, I'd also want to rename it something like `finalize` and _maybe_ have it consume `self` by value?

Reply:
So... yeah. That's also a strategy I considered. It doesn't work here because the buffer is owned by `TxMsg`. However, in light of other parts of this review, I'm questioning if `TxMsg` should be created to wrap an `&mut [u8]` and consumed for each write. That seems like the better option, and I had a version of that that I replaced that wasn't as full featured.